homeassistant/components/cover/group.py

```diff
@@ -0,0 +1,277 @@
+                               discovery_info=None):
+    """Set up the Group Cover platform."""
+    async_add_devices(
+        [CoverGroup(config[CONF_NAME], config[CONF_ENTITIES])])
+
+
+class CoverGroup(CoverDevice):
+    """Representation of a CoverGroup."""
+
+    def __init__(self, name, entities):
+        """Initialize a CoverGroup entity."""
+        self._name = name
+        self._tilt = False
+
+        self._is_closed = False
+        self._cover_position = 100
+        self._tilt_position = None
+        self._supported_features = 0
+        self._assumed_state = True
+
+        self._entities = entities
+        self._covers = {KEY_OPEN_CLOSE: set(), KEY_STOP: set(),
+                        KEY_POSITION: set()}
+        self._tilts = {KEY_OPEN_CLOSE: set(), KEY_STOP: set(),
+                       KEY_POSITION: set()}
+
+    @callback
+    def update_supported_features(self, entity_id, old_state, new_state,
+                                  update_state=True):
+        """Update dictionaries with supported features."""
+        if new_state is None:
+            for values in self._covers.values():
+                values.discard(entity_id)
+            for values in self._tilts.values():
+                values.discard(entity_id)
+        else:
+            features = new_state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+
+            # pylint: disable=expression-not-assigned
+            self._covers[KEY_OPEN_CLOSE].add(entity_id) \
+                if features & (SUPPORT_OPEN | SUPPORT_CLOSE) \
+                else self._covers[KEY_OPEN_CLOSE].discard(entity_id)
+
+            # pylint: disable=expression-not-assigned
+            self._covers[KEY_STOP].add(entity_id) \
+                if features & (SUPPORT_STOP) \
+                else self._covers[KEY_STOP].discard(entity_id)
+
+            # pylint: disable=expression-not-assigned
+            self._covers[KEY_POSITION].add(entity_id) \
+                if features & (SUPPORT_SET_POSITION) \
+                else self._covers[KEY_POSITION].discard(entity_id)
+
+            # pylint: disable=expression-not-assigned
+            self._tilts[KEY_OPEN_CLOSE].add(entity_id) \
+                if features & (SUPPORT_OPEN_TILT | SUPPORT_CLOSE_TILT) \
+                else self._tilts[KEY_OPEN_CLOSE].discard(entity_id)
+
+            # pylint: disable=expression-not-assigned
+            self._tilts[KEY_STOP].add(entity_id) \
+                if features & (SUPPORT_STOP_TILT) \
+                else self._tilts[KEY_STOP].discard(entity_id)
+
+            # pylint: disable=expression-not-assigned
+            self._tilts[KEY_POSITION].add(entity_id) \
+                if features & (SUPPORT_SET_TILT_POSITION) \
+                else self._tilts[KEY_POSITION].discard(entity_id)
+
+        tilt = False
+        for values in self._tilts.values():
+            if not values:
+                break
+        else:
+            tilt = True
+        self._tilt = tilt
```

Comment:
*Suggestion:* This too is quite hard to read and has an `all()` syntax counterpart that is much shorter and way easier to understand IMO.

```python
self._tilt = all(entities_with_feature for entities_with_feature in self._tilts.values())
```

Reply:
:+1:

Reply:
Thanks! Actually my implementation had a logic error which I didn't see. It should be `any( ... )`.